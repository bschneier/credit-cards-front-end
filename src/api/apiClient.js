import axios from 'axios';
import CONSTANTS from '../shared/constants';

class ApiClient {
  static callCreditCardsApi(method, path, callback, data, sessionId) {
    let config = { headers: { [CONSTANTS.AUTHENTICATION_HEADER] : sessionId } };
    const url = process.env.API_HOST + path;
    let apiFunc;
    switch (method) {
      case CONSTANTS.HTTP_METHODS.GET:
        config.params = data;
        apiFunc = () => { return axios.get(url, config); };
        break;
      case CONSTANTS.HTTP_METHODS.POST:
        apiFunc = () => { return axios.post(url, data, config); };
        break;
      case CONSTANTS.HTTP_METHODS.PUT:
        apiFunc = () => { return axios.put(url, data, config); };
        break;
      case CONSTANTS.HTTP_METHODS.DELETE:
        config.params = data;
        apiFunc = () => { return axios.delete(url, config); };
        break;
    }

    apiFunc().then((response) => {
      callback(response);
    }).catch((error) => {
      if(error.response.status === 401) {
        sessionStorage.removeItem(CONSTANTS.LOCAL_STORAGE.SESSION_KEY);
      }
      callback(error.response);
    });
  }
}

export default ApiClient;
