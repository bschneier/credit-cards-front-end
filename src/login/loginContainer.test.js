import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import LoginContainer from './loginContainer';
import apiClient from '../api/apiClient';
import CONSTANTS from '../shared/constants';
import Validators from '../shared/validators';

const usernameField = 'username';
const passwordField = 'password';
const usernameValue = 'myUsername';
const passwordValue = 'myPassword';
const apiMethod = CONSTANTS.HTTP_METHODS.POST;
const apiUrl = '/authenticate';

describe('LoginContainer', () => {
  test('Initial render matches snapshot', () => {
    const tree = renderer.create(<LoginContainer location={new Object()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Invalid form submissions', () => {
    describe('With no username or password', () => {
      let wrapper;

      beforeAll(() => {
        Validators.validateForm = jest.fn(Validators.validateForm);
        wrapper = mount(<LoginContainer location={new Object()} />);
        wrapper.find('form').simulate('submit');
      });

      test('Calls ValidateForm', () => {
        expect(Validators.validateForm).toHaveBeenCalledTimes(1);
      });

      test('Shows error on username field', () => {
        assertFieldErrorState(usernameField, 1, wrapper);
      });

      test('Shows error on password field', () => {
        assertFieldErrorState(passwordField, 1, wrapper);
      });
    });

    describe('With username, but no password', () => {
      let wrapper;

      beforeAll(() => {
        wrapper = mount(<LoginContainer location={new Object()} />);
        wrapper.find(`input#${usernameField}`).simulate('change',
          {target: {value: usernameValue, id: usernameField}});
        wrapper.find('form').simulate('submit');
      });

      test('Does not show error on username field', () => {
        assertFieldErrorState(usernameField, 0, wrapper);
      });

      test('Shows error on password field', () => {
        assertFieldErrorState(passwordField, 1, wrapper);
      });
    });

    describe('With password, but no username', () => {
      let wrapper;

      beforeAll(() => {
        wrapper = mount(<LoginContainer location={new Object()} />);
        wrapper.find(`input#${passwordField}`).simulate('change',
          {target: {value: passwordValue, id: passwordField}});
        wrapper.find('form').simulate('submit');
      });

      test('Shows error on username field', () => {
        assertFieldErrorState(usernameField, 1, wrapper);
      });

      test('Does not show error on password field', () => {
        assertFieldErrorState(passwordField, 0, wrapper);
      });
    });
  });

  describe('Valid form submissions', () => {
    describe('Calls API client once with correct data', () => {
      let wrapper;

      beforeEach(() => {
        Validators.validateForm = jest.fn(Validators.validateForm);
        apiClient.callCreditCardsApi = jest.fn();
        wrapper = mount(<LoginContainer location={new Object()} />);
        wrapper.find(`input#${usernameField}`).simulate('change',
          {target: {value: usernameValue, id: usernameField}});
        wrapper.find(`input#${passwordField}`).simulate('change',
          {target: {value: passwordValue, id: passwordField}});
      });

      test('Without remember me', () => {
        wrapper.find('form').simulate('submit');
        expect(Validators.validateForm).toHaveBeenCalledTimes(1);
        expect(apiClient.callCreditCardsApi).toHaveBeenCalledTimes(1);
        expect(apiClient.callCreditCardsApi).toHaveBeenCalledWith(apiMethod, apiUrl, expect.any(Function), {
          username: usernameValue,
          password: passwordValue,
          rememberMe: false
        });
      });

      test('With remember me', () => {
        wrapper.find('input[type="checkbox"]').simulate('change',
          {target: {checked: true}});
        wrapper.find('form').simulate('submit');
        expect(Validators.validateForm).toHaveBeenCalledTimes(1);
        expect(apiClient.callCreditCardsApi).toHaveBeenCalledTimes(1);
        expect(apiClient.callCreditCardsApi).toHaveBeenCalledWith(apiMethod, apiUrl, expect.any(Function), {
          username: usernameValue,
          password: passwordValue,
          rememberMe: true
        });
      });
    });
  });
});

function assertFieldErrorState(fieldName, count, wrapper) {
  let field = wrapper.find(`.has-error input#${fieldName}`);
  expect(field).toHaveLength(count);
}