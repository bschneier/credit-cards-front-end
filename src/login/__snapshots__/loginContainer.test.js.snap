// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer Initial render matches snapshot 1`] = `
<div
  className="colspan-12 center-block login-form-container"
>
  <div
    className="message-error message-hidden message-size3
    page-message"
  />
  <form
    className="form-horizontal"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="col-xs-4 control-label"
        htmlFor="username"
      >
        Username
      </label>
      <div
        className="col-xs-8"
      >
        <input
          className="credit-cards-text-input form-control"
          id="username"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="message-error message-hidden message-size2
    "
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-xs-4 control-label"
        htmlFor="password"
      >
        Password
      </label>
      <div
        className="col-xs-8"
      >
        <input
          className="credit-cards-text-input form-control"
          id="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <div
          className="message-error message-hidden message-size2
    "
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="col-xs-8 col-xs-offset-4"
      >
        <div
          className="checkbox"
          style={undefined}
        >
          <label
            title=""
          >
            <input
              checked={undefined}
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            Remember Me
          </label>
        </div>
        <div
          className="message-error message-hidden message-size2
    "
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="col-xs-8 col-xs-offset-4"
      >
        <button
          className="login-submit-button btn login-submit-button btn-default"
          disabled={false}
          onClick={undefined}
          type="submit"
        >
          Sign In
        </button>
      </div>
    </div>
  </form>
</div>
`;
